File: src/omnia/data/structure/observable/ObservableDirectedGraph.java
package omnia.data.structure.observable;

import io.reactivex.rxjava3.core.Observable;
import omnia.data.structure.DirectedGraph;
import omnia.data.structure.Set;

public interface ObservableDirectedGraph<E> extends DirectedGraph<E>, ObservableGraph<E> {
  
  @Override
  ObservableChannels<E> observe();

  interface ObservableChannels<E> extends ObservableGraph.ObservableChannels<E> {

    @Override
    Observable<? extends DirectedGraph<E>> states();

    @Override
    Observable<? extends MutationEvent<E>> mutations();
  }

  interface MutationEvent<E> extends ObservableGraph.MutationEvent<E> {

    @Override
    DirectedGraph<E> state();

    @Override
    Set<? extends ObservableGraph.GraphOperation<E>> operations();
  }
}

File: src/omnia/data/structure/observable/ObservableDirectedGraph.kt
package omnia.data.structure.observable

import com.badoo.reaktive.observable.Observable
import omnia.data.structure.DirectedGraph
import omnia.data.structure.Set
import omnia.data.structure.observable.ObservableGraph.GraphOperation

interface ObservableDirectedGraph<E : Any> : DirectedGraph<E>, ObservableGraph<E> {

  override fun observe(): ObservableChannels<E>

  interface ObservableChannels<E : Any> : ObservableGraph.ObservableChannels<E> {

    override fun states(): Observable<DirectedGraph<E>>

    override fun mutations(): Observable<MutationEvent<E>>
  }

  interface MutationEvent<E : Any> : ObservableGraph.MutationEvent<E> {

    override fun state(): DirectedGraph<E>

    override fun operations(): Set<out GraphOperation<E>>
  }
}